src/Workspaces/Core/Portable/Rename/Renamer.SyncNamespaceDocumentAction.cs

```diff
@@ -0,0 +1,114 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.ChangeNamespace;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will sync the namespace of the document to match the folders property 
+        /// of that document
+        /// </summary>
+        internal sealed class SyncNamespaceDocumentAction : RenameDocumentAction
+        {
+            private readonly AnalysisResult _analysis;
+
+            private SyncNamespaceDocumentAction(AnalysisResult analysis, ImmutableArray<ErrorResource> errors)
+                : base(errors)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+                => WorkspacesResources.ResourceManager.GetString("Sync_namespace_to_folder_structure", culture ?? WorkspacesResources.Culture)!;
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, OptionSet _, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+                var changeNamespaceService = document.GetRequiredLanguageService<IChangeNamespaceService>();
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var syntaxRoot = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                // Don't descend into anything other than top level declarations from the root.
+                // ChangeNamespaceService only controls top level declarations right now
+                var namespaceDeclarations = syntaxRoot.DescendantNodes(n => !syntaxFacts.IsDeclaration(n)).Where(n => syntaxFacts.IsNamespaceDeclaration(n));
+
+                foreach (var namespaceDeclaration in namespaceDeclarations)
+                {
+                    solution = await changeNamespaceService.ChangeNamespaceAsync(
```

Comment:
i'm also  confused about what this means if i have:

```
namespace A
{
    namespace B
    {
        class C { }
    }
}
```

and i move this from `A\B\C.cs` to `A\X\C.cs` or `Y\B\c.cs`.  we'll find both namespace A and B here... and we'll try to change both?  i don't really understand what's going on.

Reply:
Adding this test, hopefully it helps clear up some of the confusion

```csharp
        [Fact]
        public Task CSharp_RenameDocument_RenameMultipleNamespaces_Nested()
=> TestRenameDocument(
@"namespace Test.Path
{
    class C
    {
    }
}

namespace Test
{
    namespace Path
    {
        class C2
        {
        }
    }
}",
@"namespace Test.Path.After.Test
{
    class C
    {
    }
}

namespace Test
{
    namespace Path
    {
        class C2
        {
        }
    }
}",
documentPath: @"Test\Path\Document.cs",
documentName: @"Document.cs",
newDocumentPath: @"Test\Path\After\Test\Document.cs");
```

Reply:
This follows the same logic as sync namespace for now. If we want to change that behavior we should do it for both places, and as a separate PR/issue.